angular/projects/global/src/lib/services/store.service.ts

```diff
@@ -0,0 +1,153 @@
+export class StoreService {
+  private currentState: any;
+
+  constructor() {
+    this.currentState = {};
+    this.state.navigationStack = [];
+  }
+
+  get state(): any {
+    return this.currentState;
+  }
+  set state(state: any) {
+    this[state.type] = state.data;
+  }
+
+  public getNavigationStack() {
+    if (this.state.navigationStack) {
+      return this.state.navigationStack;
+    } else {
+      return null;
+    }
+  }
+
+  public getLastBackUrl() {
+    const stack = this.getNavigationStack();
+    if (stack) {
+      return stack[stack.length - 1].backUrl;
+    } else {
+      return null;
+    }
+  }
+
+  public getLastNavigationObject() {
+    const stack = this.getNavigationStack();
+    if (stack) {
+      return stack[stack.length - 1];
+    } else {
+      return null;
+    }
+  }
+
+  public pushNavigationStack(backUrl: string[], breadcrumbs: object[], componentId: string = null) {
+    const navigationObject = {
+      backUrl: backUrl,
+      breadcrumbs: breadcrumbs,
+      componentId: componentId
+    };
+    const stack = this.getNavigationStack();
+    if (stack) {
+      stack.push(navigationObject);
+      this.state.navigationStack = stack;
+    } else {
+      this.state.navigationStack = [navigationObject];
+    }
+  }
+
+  public popNavigationStack() {
+    const stack = this.getNavigationStack();
+    if (stack) {
+      let stackObject = stack.pop();
+      if (stack.length === 0) {
+        this.clearNavigationStack();
+        stackObject = null;
+      } else {
+        this.state.navigationStack = stack;
+      }
+      return stackObject;
+    } else {
+      this.clearNavigationStack();
+      return null;
+    }
+  }
+
+  public navigateBreadcrumb(breadcrumb, router) {
+    const stack = this.getNavigationStack();
+    let poppedItem = null;
+    let isPopping = true;
+    if (stack) {
+      while (isPopping) {
+        poppedItem = this.popNavigationStack();
+        if (poppedItem == null) {
+          break;
+        } else if (poppedItem.breadcrumbs[poppedItem.breadcrumbs.length - 1] === breadcrumb) {
+          isPopping = false;
+        }
+      }
+      router.navigate(breadcrumb.route);
+    } else {
+      router.navigate(['/']);
+    }
+  }
+
+  public clearNavigationStack() {
+    this.state.navigationStack = null;
+  }
+}
+
+export interface INavigationObjectParams {
+  breadcrumbs: object[];
+  componentName?: string;
+  componentId?: string;
+  pageData?: any;
+}
+
+/*
+  Example of a back Url array:
+  ['/p', this.project._id, 'edit']
+
+  Example of a breadcrumbs array:
+  Note that the objects in the array contain a route and a label.
+  [
+      {
+          route: ['/projects'],
+          label: 'All Projects'
```

Comment:
You could create a breadcrumbs type to use, instead of `object`.  Then you can enforce `route` and `label` as properties.

`breadcrumbs: IBreadcrumb[]`